### PR TITLE
[FIXIMP] core: unsuck upgrade code cli

### PR DESCRIPTION
**Pull request by Julien00859**
Moved the logs/progress on stderr so it is possible to silent them using
a shell redirection `2>/dev/null`. Simplified what is written on stdout
so it is scriptable.

Moved `sys.stdout.isatty()` outside of `print_progress` so the function
is evaluated once, and not everytime we print a progress. Should speedup
the scripts a bit on runbot.

Moved the `print_progress` in scripts inside the loop, so it doesn't try
to `print_progress(..., 0)` and crash on a division by 0. The runner now
takes care of logging 0% and 100%, along with the script name.

Changed `-t/--test` for `--dry-run`, for consistency with other posix
commands such as rsync.

Made so it exits 1 when there are dirty files.

Added a `--script` option to run a single script. Useful in development
when we want to test a single script and not run all the scripts of a
given version.

Removed the default of `--from`, we couldn't agree on a sensitive
default value. `--from` (or `--script`) is now required.

Used mature `parse_version` from the tools, which is suitable to parse
odoo versions.

Added a long text that describes what "code upgrade" is. It is used as
the module docstring and the --help description.

Used cross-plateform \r instead of unix specific \033.

Skipped `__pycache__` folders, not strickly necessary as `.pyc` files
would nevertheless be skipped via `AVAILABLE_EXT`.

Replaced some string manipulation on paths by pathlib.

### CODE REVIEW

**Comment by Julien00859** [comment]
Hello @odoo/rd-security yet again another PR where I used many *dangerous* methods.

* Inside the tests, `sp.run`, `sys.argv[0]`, `sys.executable` in order to invoke the odoo cli and test its result. The executable is odoo itself, the subcommand is forced to `upgrade_code`, we never user `shell=True` and even if we did, it's inside a test so out-of-reach of users.
* Inside `upgrade_code` itself, many path manipulation, in order to locate/discover the scripts to run. I've added a safeguard for the new `--script` option [here](<https://github.com/odoo/odoo/pull/185387/files#diff-527cbc0a9ff314cc12aafc04d0da70dff607f5cdc6a8826a7491e3e295d53620R139>) to basically *chroot* the user inside the `upgrade_code/` folder.
* Inside `upgrade_code` also, that disgusting `sys.path` manipulation, that's a request from the runbot team, they wanna be able to run the `upgrade_code` CLI by running the file directly (hence the chmod +x and the shebang) and not via *odoo-bin* in order to run the upgrade scripts with mergebot without installing all the dependencies of odoo inside the mergebot docker image. I could had used `exec()` or `importlib` to achieve the same thing put I prefer to change the syspath.

File: odoo/cli/upgrade_code.py [hunk 1]
```diff
@@ -1,40 +1,80 @@
...
+There are additional utilities on the file_manager, such as:
+* print_progress(current, total)
+
+Exemple:
+
+    def upgrade(file_manager):
+        files = [f for f in file_manager if f.path.suffix == '.py']
+        for fileno, file in enumerate(files):
+            file.content = file.content.replace(..., ...)
+            file_manager.print_progress(fileno, len(files))
```
Comment (rco-odoo) [inline odoo/cli/upgrade_code.py:23]: Couldn't we get rid of `print_progress` altogether by making the iterator of `file_manager` print the progress?  Maybe we could provide a specific iteration method on `file_manager` for that purpose...  Note that the progress is computed with respect to all the files matched by the glob, so it behaves slightly differently.  But that's a good enough approximation IMO.
```suggestion
    def upgrade(file_manager):
        for file in file_manager.iter_files():  # this iterator manages the progress
            if file.path.suffix == '.py':
                file.content = file.content.replace(..., ...)
```
I am not sure it is a good idea, but I thought it was worth sharing :wink:

**Review by rco-odoo — APPROVED** [review]
Looks like a good improvement to me.

File: odoo/cli/upgrade_code.py (hunk 1 above)
Comment (rco-odoo) [reply odoo/cli/upgrade_code.py:23]: By the way, the special iterator method could even take a parameter to filter files, and the progress could be computed with respect to the filtered files. :thinking:

Comment (Julien00859) [reply odoo/cli/upgrade_code.py:23]: I've a feeling explicit > implicit, it is a single line of code, and it is not too bad if you forget it

File: odoo/addons/base/tests/test_upgrade_code.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+from odoo.tests import BaseCase
+
+
+class TestCommand(BaseCase):
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.odoo_bin = sys.argv[0]
+        assert 'odoo-bin' in cls.odoo_bin
```
[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/addons/base/tests/test_upgrade_code.py]: done

**Review by rco-odoo — APPROVED** [review]

**Review by mart-e — APPROVED** [review]
@robodoo override=ci/security
Reason: code not callable by user but launched manually during upgrade

**Comment by Julien00859** [comment]
@robodoo r+

**Comment by Julien00859** [comment]
@robodoo retry

**Comment by Julien00859** [comment]
@robodoo r+

**Comment by moneebalalfi** [comment]
So [from, to] now are required ?

**Comment by Xavier-Do** [comment]
Yes, this will avoid confusion, better explicit than implicit.

**Pull request closed without merging**